src/v/cluster/partition_leaders_table.cc

```diff
@@ -168,6 +168,13 @@ void partition_leaders_table::update_partition_leader(
             promise.second->set_value(*leader_id);
         }
     }
+
+    // Ensure leadership has changed before notifying watchers
+    if (
+      !it->second.previous_leader
+      || leader_id.value() != it->second.previous_leader.value()) {
+        _watchers.notify(ntp, ntp, term, leader_id);
+    }
```

Comment:
tbh this looks error prone as in it might not deliver updates. for example, what if a watcher was interested in the term at which a leader changed? the conditional here will not deliver an update if leader returns to the same as previous in but in the next term. one might say that prev/curr leader change in lock step so perhaps that isn't possible, but i'm not so sure it is that strict.